Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximated_centroidal_Voronoi_diagram_remeshing.h

```diff
@@ -0,0 +1,1319 @@
+  {
+    Halfedge_descriptor hd1 = hd;
+
+    int cb_first = -1;
+
+    do
+    {
+      // 1- get the target and source vertices vt, vs
+      // 2- if the target and source vertices are in different clusters, create a new vertex vb between them vb = (vt + vs) / 2
+      // 3- make a new face with the new vertex vb and the centers of the clusters of vt and vs
+      // 4- also make a new face with vb, the next vb, and the center of the cluster of vt
+
+      Vertex_descriptor vt = target(hd1, pmesh);
+      Vertex_descriptor vs = source(hd1, pmesh);
+
+      int ct = get(vertex_cluster_pmap, vt);
+      int cs = get(vertex_cluster_pmap, vs);
+
+      if (ct != cs)
+      {
+        typename GT::Point_3 vt_p = get(vpm, vt);
+        typename GT::Point_3 vs_p = get(vpm, vs);
+        typename GT::Vector_3 vt_v(vt_p.x(), vt_p.y(), vt_p.z());
+        typename GT::Vector_3 vs_v(vs_p.x(), vs_p.y(), vs_p.z());
+
+        typename GT::Vector_3 vb_v = (vt_v + vs_v) / 2;
+        typename GT::Point_3 vb_p(vb_v.x(), vb_v.y(), vb_v.z());
+
+        points.push_back(vb_p);
+
+        std::size_t cb = points.size() - 1;
+
+        if (cb_first == -1)
+          cb_first = cb;
+
+        std::size_t ct_mapped = valid_cluster_map[ct], cs_mapped = valid_cluster_map[cs];
+
+        if (ct_mapped != std::size_t(-1) && cs_mapped != std::size_t(-1))
+        {
+          std::array<std::size_t, 3> polygon = {ct_mapped, cb, cs_mapped};
+          polygons.push_back(polygon);
+
+          // after the loop, the last cb+1 should be modified to the first cb
+          polygon = {cb, ct_mapped, cb + 1};
+          polygons.push_back(polygon);
+        }
+      }
+      hd1 = next(hd1, pmesh);
+    } while (hd1 != hd);
+    polygons[polygons.size() - 1][2] = cb_first;
+  }
+
+  // create a triangle for each face with all vertices in 3 different clusters
+  for (Face_descriptor fd : faces(pmesh))
+  {
+    Halfedge_descriptor hd1 = halfedge(fd, pmesh);
+    Vertex_descriptor v1 = source(hd1, pmesh);
+    Halfedge_descriptor hd2 = next(hd1, pmesh);
+    Vertex_descriptor v2 = source(hd2, pmesh);
+    Halfedge_descriptor hd3 = next(hd2, pmesh);
+    Vertex_descriptor v3 = source(hd3, pmesh);
+
+    int c1 = get(vertex_cluster_pmap, v1);
+    int c2 = get(vertex_cluster_pmap, v2);
+    int c3 = get(vertex_cluster_pmap, v3);
+
+    if (c1 != c2 && c1 != c3 && c2 != c3)
+    {
+      std::size_t c1_mapped = valid_cluster_map[c1], c2_mapped = valid_cluster_map[c2], c3_mapped = valid_cluster_map[c3];
+      if (c1_mapped !=  std::size_t(-1) && c2_mapped !=  std::size_t(-1) && c3_mapped !=  std::size_t(-1))
+      {
+        std::array<std::size_t,3> polygon = {c1_mapped, c2_mapped, c3_mapped};
+        polygons.push_back(polygon);
+      }
+    }
+  }
+
+static int kkk=-1;
+CGAL::IO::write_polygon_soup("/tmp/soup_"+std::to_string(++kkk)+".off", points, polygons);
+dump_mesh_with_cluster_colors(pmesh, vertex_cluster_pmap, "/tmp/cluster_"+std::to_string(kkk)+".ply");
```

Comment:
Indentation + clean + This can't be compiling with anything else than a Surface_mesh + Missing tests.